extensions/Bookalope/jsx/bookalope.jsx

```diff
@@ -193,3 +193,347 @@ function bookalopeCreateDocument(idmlFileName, bookId, bookflowId, betaHost) {
+    //
+    // @param {Graphic} graphic - The Graphic object to export.
+    // @param {File} file - The File object represents a local host file.
+    function exportPNG(graphic, file) {
+        try {
+            var other = graphic.duplicate();
+            try {
+                // FIXME Where's images documented?
+                // https://www.indesignjs.de/extendscriptAPI/indesign-latest/#Graphic.html
+                other.images[0].clearTransformations();
+                other.images[0].fit(FitOptions.FRAME_TO_CONTENT);
+            } catch (_) {
+                // Ignore any problem and try to export the image anyway.
+            }
+            other.exportFile(ExportFormat.PNG_FORMAT, file);
+            other.remove();
+        } catch (_) {
+            // TODO Notify caller/user that exporting the image failed.
+        }
+    }
+
+    // Compare the locations of two text frames based on their respective geometric
+    // boundaries (which have the format [y1, x1, y2, x2]). Returns -1 if frameA
+    // is above or left of frameB (i.e. is "smaller"); returns 1 otherwise.
+    //
+    // @param {TextFrame} frameA - A TextFrame object reference.
+    // @param {TextFrame} frameB - A TextFrame object reference.
+    // @return {Number} -1 if frameA is smaller than frameB, 1 otherwise.
+    function cmpFrames(frameA, frameB) {
+        if (frameA.geometricBounds[0] < frameB.geometricBounds[0]) {
+            return -1;
+        }
+        if (frameA.geometricBounds[0] == frameB.geometricBounds[0]) {
+            return (frameA.geometricBounds[1] < frameB.geometricBounds[1]) ? -1 : 1;
+        }
+        return 1;
+    }
+
+    // Create a temporary copy of the document that we want to export.
+    const tmpFile = File(app.createTemporaryCopy(doc.fullName))
+    const tmpDoc = app.open(tmpFile, false);
+    const tmpPath = tmpFile.path;
+
+    // Step 1: unlock all layers and elements in the document.
+    tmpDoc.layers.everyItem().locked = false;
+    tmpDoc.pageItems.everyItem().locked = false;
+
+    // Step 2: add a character style for page numbers that we'll inject
+    // into the text further down. Bookalope will know what to do with
+    // that extra goodness.
+    const pgnrCharacterStyleName = "bookalope-page-number";
+    var pgnrCharacterStyle = tmpDoc.characterStyles.itemByName(pgnrCharacterStyleName);
+    if (!pgnrCharacterStyle.isValid) {
+        pgnrCharacterStyle = tmpDoc.characterStyles.add({
+            name: pgnrCharacterStyleName,
+            pointSize: 0.1
+        });
+    }
+
+    // Step 3: insert into the text and where the text flow breaks onto the
+    // next page and using our special character style the page name of the
+    // current page.
+    var win = new Window("palette", "Processing document page numbers...");
+    win.pbar = win.add("progressbar", undefined, 0, tmpDoc.pages.length);
+    win.pbar.preferredSize.width = 300;
+    win.show();
```

Comment:
@gregoriopellegrino I like this progress bar, thank you! How about we open it when the RTF export begins and update the progress bar with every step — not just for this step 3. That way the user gets feedback for the _entire_ conversion?

Reply:
Good idea. That was part of the page references plugin, it seems to me that it might make sense to expand it to the whole process (I need to study how to do that).

Reply:
I thought that we open the progress window with a max value of `8` (because the RTF conversion has eight steps)
```javascript
    win.pbar = win.add("progressbar", undefined, 0, 8);
```
and then update the progress bar with every step
```javascript
    // Step 4: delete empty graphics, and then export the embedded images.
    win.pbar.value = 4;
```
and so on… 🤔

If that `pbar.value` can take float values then we can consider updating it during every single step, too. For example if the document has 100 pages and step 3 iterates over all pages, then with every page we can update
```javascript
    win.pbar.value += (1 / doc.pages.length);
```
which would give a more accurate update with every one of the 8 big steps.